src/test/regress/sql/multi_subquery.sql

```diff
@@ -0,0 +1,538 @@
+		a_inner AS a
+	FROM
+		(SELECT
+			subquery_pruning_varchar_test_table.a AS a_inner
+		FROM
+		 	subquery_pruning_varchar_test_table
+		GROUP BY
+		  	subquery_pruning_varchar_test_table.a
+		HAVING
+		  	count(subquery_pruning_varchar_test_table.a) < 3)
+		AS f1,
+
+		(SELECT
+		 	subquery_pruning_varchar_test_table.a
+		FROM
+		  	subquery_pruning_varchar_test_table
+		GROUP BY
+		  	subquery_pruning_varchar_test_table.a
+		HAVING
+		  	sum(coalesce(subquery_pruning_varchar_test_table.b,0)) > 20.0)
+		AS f2
+	WHERE
+		f1.a_inner = f2.a
+	GROUP BY
+		a_inner)
+AS foo;
+
+DROP TABLE subquery_pruning_varchar_test_table;
+
+-- Simple join subquery pushdown
+SELECT
+	avg(array_length(events, 1)) AS event_average
+FROM
+	(SELECT
+		tenant_id,
+		user_id,
+		array_agg(event_type ORDER BY event_time) AS events
+	FROM
+		(SELECT
+			(users.composite_id).tenant_id,
+			(users.composite_id).user_id,
+			event_type,
+			events.event_time
+		FROM
+			users,
+			events
+		WHERE
+			(users.composite_id) = (events.composite_id) AND
+			users.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
+			users.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
+			event_type IN ('click', 'submit', 'pay')) AS subquery
+	GROUP BY
+		tenant_id,
+		user_id) AS subquery;
+
+-- Union and left join subquery pushdown
+SELECT
+	avg(array_length(events, 1)) AS event_average,
+	hasdone
+FROM
+	(SELECT
+		subquery_1.tenant_id,
+		subquery_1.user_id,
+		array_agg(event ORDER BY event_time) AS events,
+		COALESCE(hasdone, 'Has not done paying') AS hasdone
+	FROM
+	(
+		(SELECT
+			(users.composite_id).tenant_id,
+			(users.composite_id).user_id,
+			(users.composite_id) as composite_id,
+			'action=>1'AS event,
+			events.event_time
+		FROM
+			users,
+			events
+		WHERE
+			(users.composite_id) = (events.composite_id) AND
+			users.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
+			users.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
+			event_type = 'click')
+		UNION
+		(SELECT
+			(users.composite_id).tenant_id,
+			(users.composite_id).user_id,
+			(users.composite_id) as composite_id,
+			'action=>2'AS event,
+			events.event_time
+		FROM
+			users,
+			events
+		WHERE
+			(users.composite_id) = (events.composite_id) AND
+			users.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
+			users.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
+			event_type = 'submit')
+	) AS subquery_1
+	LEFT JOIN
+	(SELECT
+		DISTINCT ON ((composite_id).tenant_id, (composite_id).user_id) composite_id,
+		(composite_id).tenant_id,
+		(composite_id).user_id,
+		'Has done paying'::TEXT AS hasdone
+	FROM
+		events
+	WHERE
+		events.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
+		events.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
+		event_type = 'pay') AS subquery_2
+	ON
+		subquery_1.composite_id = subquery_2.composite_id
+	GROUP BY
```

Comment:
I tried to excise all unordered queries in #1377, but it appears this change introduces quite a few more. In the future, try to find something to `ORDER` queries by so their output is deterministic (this query changes in PostgreSQL 10).